usr.sbin/tcpdump/tcpdump.8
.\"	$OpenBSD: tcpdump.8,v 1.109 2019/11/06 12:49:50 sthen Exp $
.\"
.\" Copyright (c) 1987, 1988, 1989, 1990, 1991, 1992, 1994, 1995, 1996
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that: (1) source code distributions
.\" retain the above copyright notice and this paragraph in its entirety, (2)
.\" distributions including binary code include the above copyright notice and
.\" this paragraph in its entirety in the documentation or other materials
.\" provided with the distribution, and (3) all advertising materials mentioning
.\" features or use of this software display the following acknowledgement:
.\" ``This product includes software developed by the University of California,
.\" Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
.\" the University nor the names of its contributors may be used to endorse
.\" or promote products derived from this software without specific prior
.\" written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.Dd $Mdocdate: November 6 2019 $
.Dt TCPDUMP 8
.Os
.Sh NAME
.Nm tcpdump
.Nd dump traffic on a network
.Sh SYNOPSIS
.Nm tcpdump
.Op Fl AadefILlNnOopqStvXx
.Op Fl B Ar fildrop
.Op Fl c Ar count
.Op Fl D Ar direction
.Op Fl E Oo Ar espalg : Oc Ns Ar espkey
.Op Fl F Ar file
.Op Fl i Ar interface
.Op Fl r Ar file
.Op Fl s Ar snaplen
.Op Fl T Ar type
.Op Fl w Ar file
.Op Fl y Ar datalinktype
.Op Ar expression
.Sh DESCRIPTION
.Nm
prints out the headers of packets on a network interface that match the boolean
.Ar expression .
You must have read access to
.Pa /dev/bpf .
.Pp
The options are as follows:
.Bl -tag -width "-c count"
.It Fl A
Print each packet in ASCII.
If the
.Fl e
option is also specified, the link-level header will be included.
The smaller of the entire packet or
.Ar snaplen
bytes will be printed.
.It Fl a
Attempt to convert network and broadcast addresses to names.
.It Fl B Ar fildrop
Configure the drop action specified by
.Ar fildrop
to be used when the filter expression matches a packet.
The actions are:
.Pp
.Bl -tag -width "capture" -offset indent -compact
.It Cm pass
Matching packets are accepted and captured.
.It Cm capture
Matching packets are dropped and captured.
.It Cm drop
Matching packets are dropped but not captured.
.El
.Pp
The default action is
.Cm pass .
.It Fl c Ar count
Exit after receiving
.Ar count
packets.
.It Fl D Ar direction
Select packets flowing in the specified
.Ar direction .
Valid directions are:
.Cm in
and
.Cm out .
The default is to accept packets flowing in any direction.
.It Fl d
Dump the compiled packet-matching code in a human readable form to
standard output and stop.
.It Fl dd
Dump packet-matching code as a C program fragment.
.It Fl ddd
Dump packet-matching code as decimal numbers
preceded with a count.
.It Fl E Oo Ar espalg : Oc Ns Ar espkey
Try to decrypt RFC 4835 ESP
.Pq Encapsulating Security Payload
traffic using the specified hex key
.Ar espkey .
Supported algorithms for
.Ar espalg
are:
.Cm aes128 ,
.Cm aes128-hmac96 ,
.Cm blowfish ,
.Cm blowfish-hmac96 ,
.Cm cast ,
.Cm cast-hmac96 ,
.Cm des3 ,
.Cm des3-hmac96 ,
.Cm des
and
.Cm des-hmac96 .
The algorithm defaults to
.Cm aes128-hmac96 .
This option should be used for debugging only, since the key will show up in
.Xr ps 1
output.
.It Fl e
Print the link-level header on each dump line.
.It Fl F Ar file
Use
.Ar file
as input for the filter expression.
Any additional expressions given on the command line are ignored.
.It Fl f
Print
.Dq foreign
internet addresses numerically rather than symbolically.
This option is intended to get around serious brain damage in
Sun's yp server \(em usually it hangs forever translating non-local
internet numbers.
.It Fl I
Print the interface on each dump line.
.It Fl i Ar interface
Listen on
.Ar interface .
If unspecified,
.Nm
searches the system interface list for the lowest numbered, configured
.Dq up
interface
.Pq excluding loopback .
Ties are broken by choosing the earliest match.
.Ar interface
may be either a network interface or a USB interface, for example
.Ar usb0 .
.It Fl L
List the supported data link types for the interface and exit.
.It Fl l
Make stdout line buffered.
Useful if you want to see the data while capturing it.
For example:
.Pp
.Dl # tcpdump -l | tee dat
or
.Dl # tcpdump -l > dat & tail -f dat
.It Fl N
Do not print domain name qualification of host names.
For example, if you specify this flag then
.Nm
will print
.Dq nic
instead of
.Dq nic.ddn.mil .
.It Fl n
Do not convert addresses
.Pq host addresses, port numbers, etc.
to names.
.It Fl O
Do not run the packet-matching code optimizer.
This is useful only if you suspect a bug in the optimizer.
.It Fl o
Print a guess of the possible operating system(s) of hosts that sent
TCP SYN packets.
See
.Xr pf.os 5
for a description of the passive operating system fingerprints.
.It Fl p
Do not put the interface into promiscuous mode.
The interface might be in promiscuous mode for some other reason; hence,
.Fl p
cannot be used as an abbreviation for
.Dq ether host \&"{local-hw-addr}\&"
or
.Dq ether broadcast .
.It Fl q
Quick
.Pq quiet?
output.
Print less protocol information so output lines are shorter.
.It Fl r Ar file
Read packets from a
.Ar file
which was created with the
.Fl w
option.
Standard input is used if
.Ar file
is
.Ql - .
.It Fl S
Print absolute, rather than relative, TCP sequence numbers.
.It Fl s Ar snaplen
Analyze at most the first
.Ar snaplen
bytes of data from each packet rather than the default of 116.
116 bytes is adequate for IPv6, ICMP, TCP, and UDP,
but may truncate protocol information from name server and NFS packets
.Pq see below .
Packets truncated because of a limited
.Ar snaplen
are indicated in the output with
.Dq Op | Ns Em proto ,
where
.Em proto
is the name of the protocol level at which the truncation has occurred.
Taking larger snapshots both increases the amount of time it takes
to process packets and, effectively, decreases the amount of packet buffering.
This may cause packets to be lost.
You should limit
.Ar snaplen
to the smallest number that will capture the protocol information
you're interested in.
.It Fl T Ar type
Force packets selected by
.Ar expression
to be interpreted as the specified
.Ar type .
Currently known types are:
.Pp
.Bl -tag -width "erspan" -offset indent -compact
.It Cm cnfp
Cisco NetFlow protocol
.It Cm erspan
Cisco Encapsulated Remote Switch Port Analyzer (ERSPAN) over GRE
.It Cm gre
Generic Routing Encapsulation over UDP
.It Cm mpls
Multiprocol Label Switching over UDP
.It Cm rpc
Remote Procedure Call
.It Cm rtcp
Real-Time Applications control protocol
.It Cm rtp
Real-Time Applications protocol
.It Cm sack
RFC 2018 TCP Selective Acknowledgements Options
.It Cm tcp
Transmission Control Protocol
.It Cm tftp
Trivial File Transfer Protocol
.It Cm vat
Visual Audio Tool
.It Cm vrrp
Virtual Router Redundancy protocol
.It Cm vxlan
Virtual eXtensible Local Area Network
.It Cm wb
distributed White Board
.It Cm wg
WireGuard VPN protocol
.El
.It Fl t
Do not print a timestamp on each dump line.
.It Fl tt
Print an unformatted timestamp on each dump line.
.It Fl ttt
Print day and month in timestamp.
.It Fl tttt
Print timestamp difference between packets.
.It Fl ttttt
Print timestamp difference since the first packet.
.It Fl v
.Pq Slightly more
verbose output.
For example, the time to live
.Pq TTL
and type of service
.Pq ToS
information in an IP packet are printed.
.It Fl vv
Even more verbose output.
For example, additional fields are printed from NFS reply packets.
.It Fl w Ar file
Write the raw packets to
.Ar file
rather than parsing and printing them out.
They can be analyzed later with the
.Fl r
option.
Standard output is used if
.Ar file
is
.Ql - .
.It Fl X
Print each packet in hex and ASCII.
If the
.Fl e
option is also specified, the link-level header will be included.
The smaller of the entire packet or
.Ar snaplen
bytes will be printed.
.It Fl x
Print each packet in hex.
If the
.Fl e
option is also specified, the link-level header will be included.
The smaller of the entire packet or
.Ar snaplen
bytes will be printed.
.It Fl y Ar datalinktype
Set the data link type to use while capturing to
.Ar datalinktype .
Commonly used types include
.Cm EN10MB ,
.Cm IEEE802_11 ,
and
.Cm IEEE802_11_RADIO .
The choices applicable to a particular device can be listed using
.Fl L .
.El
.Pp
.Ar expression
selects which packets will be dumped.
If no
.Ar expression
is given, all packets on the net will be dumped.
Otherwise, only packets satisfying
.Ar expression
will be dumped.
.Pp
The
.Ar expression
consists of one or more primitives.
Primitives usually consist of an
.Ar id
.Pq name or number
preceded by one or more qualifiers.
There are three different kinds of qualifiers:
.Bl -tag -width "proto"
.It Ar type
Specify which kind of address component the
.Ar id
name or number refers to.
Possible types are
.Cm host ,
.Cm net
and
.Cm port .
E.g.,
.Dq host foo ,
.Dq net 128.3 ,
.Dq port 20 .
If there is no type qualifier,
.Cm host
is assumed.
.It Ar dir
Specify a particular transfer direction to and/or from
.Ar id .
Possible directions are
.Cm src ,
.Cm dst ,
.Cm src or dst ,
.Cm src and dst ,
.Cm addr1 ,
.Cm addr2 ,
.Cm addr3 ,
and
.Cm addr4 .
E.g.,
.Dq src foo ,
.Dq dst net 128.3 ,
.Dq src or dst port ftp-data .
If there is no
.Ar dir
qualifier,
.Cm src or dst
is assumed.
The
.Cm addr1 ,
.Cm addr2 ,
.Cm addr3 ,
and
.Cm addr4
qualifiers are only valid for IEEE 802.11 Wireless LAN link layers.
For null link layers (i.e., point-to-point protocols such as SLIP
.Pq Serial Line Internet Protocol
or the
.Xr pflog 4
header), the
.Cm inbound
and
.Cm outbound
qualifiers can be used to specify a desired direction.
.It Ar proto
Restrict the match to a particular protocol.
Possible protocols are:
.Cm ah ,
.Cm arp ,
.Cm atalk ,
.Cm decnet ,
.Cm esp ,
.Cm ether ,
.Cm fddi ,
.Cm icmp ,
.Cm icmp6 ,
.Cm igmp ,
.Cm igrp ,
.Cm ip ,
.Cm ip6 ,
.Cm lat ,
.Cm mopdl ,
.Cm moprc ,
.Cm pim ,
.Cm rarp ,
.Cm sca ,
.Cm stp ,
.Cm tcp ,
.Cm udp ,
and
.Cm wlan .
E.g.,
.Dq ether src foo ,
.Dq arp net 128.3 ,
.Dq tcp port 21 ,
.Dq wlan addr1 0:2:3:4:5:6 .
If there is no protocol qualifier,
all protocols consistent with the type are assumed.
E.g.,
.Dq src foo
means
.Do
.Pq ip or arp or rarp
src foo
.Dc
.Pq except the latter is not legal syntax ;
.Dq net bar
means
.Do
.Pq ip or arp or rarp
net bar
.Dc ;
and
.Dq port 53
means
.Do
.Pq TCP or UDP
port 53
.Dc .
.Pp
.Cm fddi
is actually an alias for
.Cm ether ;
the parser treats them identically as meaning
.Qo
the data link level used on the specified network interface
.Qc .
FDDI
.Pq Fiber Distributed Data Interface
headers contain Ethernet-like source and destination addresses,
and often contain Ethernet-like packet types,
so you can filter on these FDDI fields just as with the analogous
Ethernet fields.
FDDI headers also contain other fields,
but you cannot name them explicitly in a filter expression.
.El
.Pp
In addition to the above, there are some special primitive
keywords that don't follow the pattern:
.Cm gateway ,
.Cm broadcast ,
.Cm less ,
.Cm greater ,
and arithmetic expressions.
All of these are described below.
.Pp
More complex filter expressions are built up by using the words
.Cm and ,
.Cm or ,
and
.Cm not
to combine primitives
e.g.,
.Do
host foo and not port ftp and not port ftp-data
.Dc .
To save typing, identical qualifier lists can be omitted
e.g.,
.Dq tcp dst port ftp or ftp-data or domain
is exactly the same as
.Do
tcp dst port ftp or tcp dst port ftp-data or tcp dst port domain
.Dc .
.Pp
Allowable primitives are:
.Bl -tag -width "ether proto proto"
.It Cm dst host Ar host
True if the IP destination field of the packet is
.Ar host ,
which may be either an address or a name.
.It Cm src host Ar host
True if the IP source field of the packet is
.Ar host .
.It Cm host Ar host
True if either the IP source or destination of the packet is
.Ar host .
.Pp
Any of the above
.Ar host
expressions can be prepended with the keywords,
.Cm ip ,
.Cm arp ,
or
.Cm rarp
as in:
.Pp
.D1 Cm ip host Ar host
.Pp
which is equivalent to:
.Bd -ragged -offset indent
.Cm ether proto
.Ar ip
.Cm and host
.Ar host
.Ed
.Pp
If
.Ar host
is a name with multiple IP addresses, each address will be checked for a match.
.It Cm ether dst Ar ehost
True if the Ethernet destination address is
.Ar ehost .
.Ar ehost
may be either a name from
.Pa /etc/ethers
or a number (see
.Xr ether_aton 3
for a numeric format).
.It Cm ether src Ar ehost
True if the Ethernet source address is
.Ar ehost .
.It Cm ether host Ar ehost
True if either the Ethernet source or destination address is
.Ar ehost .
.It Cm gateway Ar host
True if the packet used
.Ar host
as a gateway; i.e., the Ethernet source or destination address was
.Ar host
but neither the IP source nor the IP destination was
.Ar host .
.Ar host
must be a name and must be found in both
.Pa /etc/hosts
and
.Pa /etc/ethers .
An equivalent expression is
.Bd -ragged -offset indent
.Cm ether host
.Ar ehost
.Cm and not host
.Ar host
.Ed
.Pp

usr.sbin/tcpdump/tcpdump.c
/*	$OpenBSD: tcpdump.c,v 1.92 2020/01/24 22:46:37 procter Exp $	*/

/*
 * Copyright (c) 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that: (1) source code distributions
 * retain the above copyright notice and this paragraph in its entirety, (2)
 * distributions including binary code include the above copyright notice and
 * this paragraph in its entirety in the documentation or other materials
 * provided with the distribution, and (3) all advertising materials mentioning
 * features or use of this software display the following acknowledgement:
 * ``This product includes software developed by the University of California,
 * Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
 * the University nor the names of its contributors may be used to endorse
 * or promote products derived from this software without specific prior
 * written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

/*
 * tcpdump - monitor tcp/ip traffic on an ethernet.
 *
 * First written in 1987 by Van Jacobson, Lawrence Berkeley Laboratory.
 * Mercilessly hacked and occasionally improved since then via the
 * combined efforts of Van, Steve McCanne and Craig Leres of LBL.
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <sys/wait.h>

#include <netinet/in.h>

#include <pcap.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>

#include "interface.h"
#include "addrtoname.h"
#include "setsignal.h"
#include "gmt2local.h"

#include <sys/socket.h>
#include <net/if.h>
#include <net/pfvar.h>
#include "pfctl.h"
#include "pfctl_parser.h"
#include "privsep.h"

int Aflag;			/* dump ascii */
int aflag;			/* translate network and broadcast addresses */
int Bflag;			/* BPF fildrop setting */
int dflag;			/* print filter code */
int eflag;			/* print ethernet header */
int fflag;			/* don't translate "foreign" IP address */
int Iflag;			/* include interface in output */
int Lflag;			/* List available link types */
int nflag;			/* leave addresses as numbers */
int Nflag;			/* remove domains from printed host names */
int Oflag = 1;			/* run filter code optimizer */
int oflag;			/* print passive OS fingerprints */
int pflag;			/* don't go promiscuous */
int qflag;			/* quick (shorter) output */
int Sflag;			/* print raw TCP sequence numbers */
int tflag = 1;			/* print packet arrival time */
int vflag;			/* verbose */
int xflag;			/* print packet in hex */
int Xflag;			/* print packet in emacs-hexl style */

int packettype;

char *program_name;
char *device = NULL;

int32_t thiszone;		/* seconds offset from gmt to local time */

extern volatile pid_t child_pid;

/* Externs */
extern void bpf_dump(struct bpf_program *, int);
extern int esp_init(char *);

/* Forwards */
void	cleanup(int);
void	gotchld(int);
extern __dead void usage(void);

/* Length of saved portion of packet. */
int snaplen = 0;

struct printer {
	pcap_handler f;
	int type;
};

/* XXX needed if using old bpf.h */
#ifndef DLT_ATM_RFC1483
#define DLT_ATM_RFC1483 11
#endif

static struct printer printers[] = {
	{ ether_if_print,		DLT_EN10MB },
	{ ether_if_print,		DLT_IEEE802 },
	{ sl_if_print,			DLT_SLIP },
	{ sl_bsdos_if_print,		DLT_SLIP_BSDOS },
	{ ppp_if_print,			DLT_PPP },
	{ ppp_hdlc_if_print,		DLT_PPP_SERIAL },
	{ fddi_if_print,		DLT_FDDI },
	{ null_if_print,		DLT_NULL },
	{ raw_if_print,			DLT_RAW },
	{ atm_if_print,			DLT_ATM_RFC1483 },
	{ loop_if_print,		DLT_LOOP },
	{ enc_if_print,			DLT_ENC },
	{ pflog_if_print,		DLT_PFLOG },
	{ pfsync_if_print,		DLT_PFSYNC },
	{ ppp_ether_if_print,		DLT_PPP_ETHER },
	{ ieee802_11_if_print,		DLT_IEEE802_11 },
	{ ieee802_11_radio_if_print,	DLT_IEEE802_11_RADIO },
	{ ofp_if_print,			DLT_OPENFLOW },
	{ usbpcap_if_print,		DLT_USBPCAP },
	{ NULL,				0 },
};

static pcap_handler
lookup_printer(int type)
{
	struct printer *p;

	for (p = printers; p->f; ++p) {
		if (type == p->type)
			return p->f;
	}

	error("unknown data link type 0x%x", type);
	/* NOTREACHED */
}

static int
init_pfosfp(void)
{
	pf_osfp_initialize();
	if (pfctl_file_fingerprints(-1,
	    PF_OPT_QUIET|PF_OPT_NOACTION, PF_OSFP_FILE) == 0)
		return 1;
	return 0;
}

static pcap_t *pd;

/* Multiple DLT support */
void		 pcap_list_linktypes(pcap_t *);
void		 pcap_print_linktype(u_int);

void
pcap_print_linktype(u_int dlt)
{
	const char *name;

	if ((name = pcap_datalink_val_to_name(dlt)) != NULL)
		fprintf(stderr, "%s\n", name);
	else
		fprintf(stderr, "<unknown: %u>\n", dlt);
}

void
pcap_list_linktypes(pcap_t *p)
{
	int fd = p->fd;
	u_int n;

#define MAXDLT	100

	u_int dltlist[MAXDLT];
	struct bpf_dltlist dl = {MAXDLT, dltlist};

	if (fd < 0)
		error("Invalid bpf descriptor");

	if (ioctl(fd, BIOCGDLTLIST, &dl) == -1)
		err(1, "BIOCGDLTLIST");

	if (dl.bfl_len > MAXDLT)
		error("Invalid number of linktypes: %u", dl.bfl_len);

	fprintf(stderr, "%d link type%s supported:\n", dl.bfl_len,
	    dl.bfl_len == 1 ? "" : "s");

	for (n = 0; n < dl.bfl_len; n++) {
		fprintf(stderr, "\t");
		pcap_print_linktype(dltlist[n]);
	}
}

int
main(int argc, char **argv)
{
	int cnt = -1, op, i;
	bpf_u_int32 localnet, netmask;
	char *cp, *infile = NULL, *RFileName = NULL;
	char ebuf[PCAP_ERRBUF_SIZE], *WFileName = NULL;
	pcap_handler printer;
	struct bpf_program *fcode;
	u_char *pcap_userdata;
	u_int dirfilt = 0, dlt = (u_int) -1;
	const char *errstr;

	if ((cp = strrchr(argv[0], '/')) != NULL)
		program_name = cp + 1;
	else
		program_name = argv[0];

	/* '-P' used internally, exec privileged portion */
	if (argc >= 2 && strcmp("-P", argv[1]) == 0)
		priv_exec(argc, argv);

	if (priv_init(argc, argv))
		error("Failed to setup privsep");

	/* state: STATE_INIT */

	opterr = 0;
	while ((op = getopt(argc, argv,
	    "AaB:c:D:deE:fF:i:IlLnNOopqr:s:StT:vw:xXy:Y")) != -1)
		switch (op) {

		case 'A':
			xflag = 1;
			Aflag = 1;
			break;

		case 'a':
			aflag = 1;
			break;

		case 'B':
			if (strcasecmp(optarg, "pass") == 0)
				Bflag = BPF_FILDROP_PASS;
			else if (strcasecmp(optarg, "capture") == 0)
				Bflag = BPF_FILDROP_CAPTURE;
			else if (strcasecmp(optarg, "drop") == 0)
				Bflag = BPF_FILDROP_DROP;
			else {
				error("invalid BPF fildrop option: %s",
				    optarg);
			}
			break;

		case 'c':
			cnt = strtonum(optarg, 1, INT_MAX, &errstr);
			if (errstr)
				error("invalid packet count %s: %s",
				    optarg, errstr);
			break;

		case 'D':
			if (strcasecmp(optarg, "in") == 0)
				dirfilt = BPF_DIRECTION_OUT;
			else if (strcasecmp(optarg, "out") == 0)
				dirfilt = BPF_DIRECTION_IN;
			else
				error("invalid traffic direction %s", optarg);
			break;

		case 'd':
			++dflag;
			break;
		case 'e':
			eflag = 1;
			break;

		case 'f':
			fflag = 1;
			break;

		case 'F':
			infile = optarg;
			break;

		case 'i':
			device = optarg;
			break;

		case 'I':
			Iflag = 1;
			break;

		case 'l':
			setvbuf(stdout, NULL, _IOLBF, 0);
			break;
		case 'L':
			Lflag = 1;
			break;
		case 'n':
			nflag = 1;
			break;

		case 'N':
			Nflag = 1;
			break;

		case 'O':
			Oflag = 0;
			break;

		case 'o':
			oflag = 1;
			break;

		case 'p':
			pflag = 1;
			break;

		case 'q':
			qflag = 1;
			break;

		case 'r':
			RFileName = optarg;
			break;

		case 's':
			snaplen = strtonum(optarg, 1, INT_MAX, &errstr);
			if (errstr)
				error("invalid snaplen %s: %s", optarg, errstr);
			break;

		case 'S':
			Sflag = 1;
			break;

		case 't':
			--tflag;
			break;

		case 'T':
			if (strcasecmp(optarg, "vat") == 0)
				packettype = PT_VAT;
			else if (strcasecmp(optarg, "wb") == 0)
				packettype = PT_WB;
			else if (strcasecmp(optarg, "rpc") == 0)
				packettype = PT_RPC;
			else if (strcasecmp(optarg, "rtp") == 0)
				packettype = PT_RTP;
			else if (strcasecmp(optarg, "rtcp") == 0)
				packettype = PT_RTCP;
			else if (strcasecmp(optarg, "cnfp") == 0)
				packettype = PT_CNFP;
			else if (strcasecmp(optarg, "vrrp") == 0)
				packettype = PT_VRRP;
			else if (strcasecmp(optarg, "tcp") == 0)
				packettype = PT_TCP;
			else if (strcasecmp(optarg, "gre") == 0)
				packettype = PT_GRE;
			else if (strcasecmp(optarg, "vxlan") == 0)
				packettype = PT_VXLAN;
			else if (strcasecmp(optarg, "erspan") == 0)
				packettype = PT_ERSPAN;
			else if (strcasecmp(optarg, "mpls") == 0)
				packettype = PT_MPLS;
			else if (strcasecmp(optarg, "tftp") == 0)
				packettype = PT_TFTP;
			else if (strcasecmp(optarg, "wg") == 0)
				packettype = PT_WG;
			else if (strcasecmp(optarg, "sack") == 0)
				/*
				 * kept for compatibility; DEFAULT_SNAPLEN
				 * used to be too short to capture SACK.
				 */
				;
			else
				error("unknown packet type `%s'", optarg);
			break;

		case 'v':
			++vflag;
			break;

		case 'w':
			WFileName = optarg;
			break;
#ifdef YYDEBUG
		case 'Y':
			{
			/* Undocumented flag */
			extern int yydebug;
			yydebug = 1;
			}
			break;
#endif
		case 'y':
			i = pcap_datalink_name_to_val(optarg);
			if (i < 0)
				error("invalid data link type: %s", optarg);
			dlt = (u_int)i;
			break;

		case 'x':
			xflag = 1;
			break;

		case 'X':
			Xflag = 1;
			xflag = 1;
			break;

		case 'E':
			if (esp_init(optarg) < 0)
				error("bad esp specification `%s'", optarg);
			break;

		default:
			usage();
			/* NOTREACHED */
		}

	if (snaplen == 0) {
		switch (dlt) {
		case DLT_IEEE802_11:
			snaplen = IEEE802_11_SNAPLEN;
			break;
		case DLT_IEEE802_11_RADIO:
			snaplen = IEEE802_11_RADIO_SNAPLEN;
			break;
		default:
			snaplen = DEFAULT_SNAPLEN;
			break;
		}
	}

	if (aflag && nflag)
		error("-a and -n options are incompatible");

	if (RFileName != NULL) {
		pd = priv_pcap_offline(RFileName, ebuf);
		if (pd == NULL)
			error("%s", ebuf);
		/* state: STATE_BPF */
		localnet = 0;
		netmask = 0;
		if (fflag != 0)
			error("-f and -r options are incompatible");
	} else {
		if (device == NULL) {
			device = pcap_lookupdev(ebuf);
			if (device == NULL)
				error("%s", ebuf);
		}
		pd = priv_pcap_live(device, snaplen, !pflag, 1000, ebuf,
		    dlt, dirfilt, Bflag);
		if (pd == NULL)
			error("%s", ebuf);

		/* state: STATE_BPF */
		if (pcap_lookupnet(device, &localnet, &netmask, ebuf)) {
			if (fflag)
				warning("%s", ebuf);
			localnet = 0;
			netmask = 0;
		}
	}
	i = pcap_snapshot(pd);
	if (snaplen < i) {
		warning("snaplen raised from %d to %d", snaplen, i);
		snaplen = i;
	}

	if (Lflag) {
		pcap_list_linktypes(pd);
		exit(0);
	}

	fcode = priv_pcap_setfilter(pd, Oflag, netmask);
	/* state: STATE_FILTER */
	if (fcode == NULL)
		error("%s", pcap_geterr(pd));
	if (dflag) {
		bpf_dump(fcode, dflag);
		exit(0);
	}
	if (oflag)
		oflag = init_pfosfp();
	init_addrtoname(localnet, netmask);

	if (WFileName) {
		pcap_dumper_t *p;

		p = priv_pcap_dump_open(pd, WFileName);
		/* state: STATE_RUN */
		if (p == NULL)
			error("%s", pcap_geterr(pd));
		{
			FILE *fp = (FILE *)p;	/* XXX touching pcap guts! */
			fflush(fp);
			setvbuf(fp, NULL, _IONBF, 0);
		}
		printer = pcap_dump;
		pcap_userdata = (u_char *)p;
	} else {
		printer = lookup_printer(pcap_datalink(pd));
		pcap_userdata = NULL;
		priv_init_done();
		/* state: STATE_RUN */
	}
	if (RFileName == NULL) {
		(void)fprintf(stderr, "%s: listening on %s, link-type ",
		    program_name, device);
		pcap_print_linktype(pd->linktype);
		(void)fflush(stderr);
	}

	if (tflag > 0)
		thiszone = gmt2local(0);

	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");

	if (pcap_loop(pd, cnt, printer, pcap_userdata) < 0) {
		(void)fprintf(stderr, "%s: pcap_loop: %s\n",
		    program_name, pcap_geterr(pd));
		exit(1);
	}
	pcap_close(pd);
	exit(0);
}

/* make a clean exit on interrupts */
void
cleanup(int signo)
{
	struct pcap_stat stat;
	sigset_t allsigs;

	sigfillset(&allsigs);
	sigprocmask(SIG_BLOCK, &allsigs, NULL);

	/* Can't print the summary if reading from a savefile */
	dprintf(STDERR_FILENO, "\n");
	if (pd != NULL && pcap_file(pd) == NULL) {
		if (priv_pcap_stats(&stat) < 0) {
			dprintf(STDERR_FILENO,
			    "pcap_stats: %s\n", pcap_geterr(pd));
		} else {
			dprintf(STDERR_FILENO,
			    "%u packets received by filter\n", stat.ps_recv);
			dprintf(STDERR_FILENO,
			    "%u packets dropped by kernel\n", stat.ps_drop);
		}
	}
	_exit(0);
}

void
gotchld(int signo)
{
	pid_t pid;
	int status;
	int save_err = errno;

	do {
		pid = waitpid(child_pid, &status, WNOHANG);
		if (pid > 0 && (WIFEXITED(status) || WIFSIGNALED(status)))
			cleanup(0);
	} while (pid == -1 && errno == EINTR);

	if (pid == -1)
		_exit(1);

	errno = save_err;
}

/* dump the buffer in `emacs-hexl' style */
void
default_print_hexl(const u_char *cp, unsigned int length)
{
	unsigned int i, j, jm;
	int c;
	char ln[128], buf[128];

	printf("\n");
	for (i = 0; i < length; i += 0x10) {
		snprintf(ln, sizeof(ln), "  %04x: ", (unsigned int)i);
		jm = length - i;
		jm = jm > 16 ? 16 : jm;

		for (j = 0; j < jm; j++) {
			if ((j % 2) == 1)
				snprintf(buf, sizeof(buf), "%02x ",
				    (unsigned int)cp[i+j]);
			else
				snprintf(buf, sizeof(buf), "%02x",
				    (unsigned int)cp[i+j]);
			strlcat(ln, buf, sizeof ln);
		}
		for (; j < 16; j++) {
			if ((j % 2) == 1)
				snprintf(buf, sizeof buf, "   ");
			else
				snprintf(buf, sizeof buf, "  ");
			strlcat(ln, buf, sizeof ln);
		}

		strlcat(ln, " ", sizeof ln);
		for (j = 0; j < jm; j++) {
			c = cp[i+j];
			c = isprint(c) ? c : '.';
			buf[0] = c;
			buf[1] = '\0';
			strlcat(ln, buf, sizeof ln);
		}
		printf("%s\n", ln);
	}
}

/* dump the text from the buffer */
void
default_print_ascii(const u_char *cp, unsigned int length)
{
	int c, i;

	printf("\n");
	for (i = 0; i < length; i++) {
		c = cp[i];
		if (isprint(c) || c == '\t' || c == '\n' || c == '\r')
			putchar(c);
		else
			putchar('.');
	}
}

/* Like default_print() but data need not be aligned */
void
default_print_unaligned(const u_char *cp, u_int length)
{
	u_int i, s;
	int nshorts;

	if (Xflag) {
		/* dump the buffer in `emacs-hexl' style */
		default_print_hexl(cp, length);
	} else if (Aflag) {
		/* dump the text in the buffer */
		default_print_ascii(cp, length);
	} else {
		/* dump the buffer in old tcpdump style */
		nshorts = (u_int) length / sizeof(u_short);
		i = 0;
		while (--nshorts >= 0) {
			if ((i++ % 8) == 0)
				printf("\n\t\t\t");
			s = *cp++;
			printf(" %02x%02x", s, *cp++);
		}
		if (length & 1) {
			if ((i % 8) == 0)
				printf("\n\t\t\t");
			printf(" %02x", *cp);
		}
	}
}

void
default_print(const u_char *bp, u_int length)
{
	const u_short *sp;
	u_int i;
	int nshorts;

	if (Xflag) {
		/* dump the buffer in `emacs-hexl' style */
		default_print_hexl(bp, length);
	} else if (Aflag) {
		/* dump the text in the buffer */
		default_print_ascii(bp, length);
	} else {
		/* dump the buffer in old tcpdump style */
		if ((long)bp & 1) {
			default_print_unaligned(bp, length);
			return;
		}
		sp = (u_short *)bp;
		nshorts = (u_int) length / sizeof(u_short);
		i = 0;
		while (--nshorts >= 0) {
			if ((i++ % 8) == 0)
				printf("\n\t\t\t");
			printf(" %04x", ntohs(*sp++));
		}
		if (length & 1) {
			if ((i % 8) == 0)
				printf("\n\t\t\t");
			printf(" %02x", *(u_char *)sp);
		}
	}
}

void
set_slave_signals(void)
{
	setsignal(SIGTERM, cleanup);
	setsignal(SIGINT, cleanup);
	setsignal(SIGCHLD, gotchld);
	setsignal(SIGHUP, cleanup);
}

__dead void
usage(void)
{
	(void)fprintf(stderr,
"Usage: %s [-AadefILlNnOopqStvXx] [-B fildrop] [-c count] [-D direction]\n",
	    program_name);
	(void)fprintf(stderr,
"\t       [-E [espalg:]espkey] [-F file] [-i interface] [-r file]\n");
	(void)fprintf(stderr,
"\t       [-s snaplen] [-T type] [-w file] [-y datalinktype] [expression]\n");
	exit(1);
}
